diff --git a/src/gui/CMakeLists.txt b/src/gui/CMakeLists.txt
index 6e4acf203..e1bb67ba4 100644
--- a/src/gui/CMakeLists.txt
+++ b/src/gui/CMakeLists.txt
@@ -1,7 +1,7 @@
 project(gui)
 find_package(Qt5 REQUIRED COMPONENTS Widgets Svg Qml Quick QuickControls2 QuickWidgets Xml Network)
 find_package(KF5Archive REQUIRED)
-find_package(KF5GuiAddons)
+find_package(KF5GuiAddons 5.103)
 
 if(QUICK_COMPILER)
         find_package(Qt5QuickCompiler)
